### PR TITLE
Feature belongsToMany mixed data

### PR DESCRIPTION
**Pull request by cjquinn**
Added ability to create new belongsToMany associations where the data array is a mix of new and existing associated data.

### CODE REVIEW

**Comment by markstory** [comment]
You have [a few PHPCS errors](https://travis-ci.org/cakephp/cakephp/jobs/56915128)

File: src/ORM/Marshaller.php [hunk 1]
```diff
@@ -276,20 +276,26 @@ protected function _belongsToMany(Association $assoc, array $data, $options = []
...
+                if (!isset($query)) {
+                    $primaryCount = count($primaryKey);
+                    $query = $assoc->find();
+                }
                 $keys = array_intersect_key($row, $primaryKey);
                 if (count($keys) === $primaryCount) {
                     $query->orWhere($keys);
                 }
+            } else {
+                $records = array_merge($records, $this->many([$row], $options));
```
Comment (markstory) [inline src/ORM/Marshaller.php]: You can just call one().

File: tests/TestCase/ORM/MarshallerTest.php [hunk 2]
```diff
@@ -558,6 +558,51 @@ public function testOneBelongsToManyJoinDataAssociatedWithIds()
...
+                'Tags'
+            ]
+        ]);
+
+        $this->assertEquals($article->tags[2], $tags->get(1));
+
+        $tag_count = $tags->find()->count();
+        $articles->save($article);
+
+        $this->assertEquals($tag_count + 2, $tags->find()->count());
```
Comment (markstory) [inline tests/TestCase/ORM/MarshallerTest.php]: You should also assert that indexes 0, 1 marshalled correctly and are marked as isNew().

**Comment by lorenzo** [comment]
Thanks @cjquinn, great job :clap:

**Comment by cjquinn** [comment]
@lorenzo thanks! Should I open an issue prior to a pull request to discuss in the future?

**Comment by markstory** [comment]
@cjquinn Nope, we're quite happy to discuss working code. Issues are good if you want to get feedback before writing any code.

**Comment by markstory** [comment]
This should be added to the docs. I will make an issue for that.

**Pull request merged**